// @redwoodjs/cli/dist/commands/dbCommands/introspect.js
"use strict";

var _interopRequireDefault = require("@babel/runtime-corejs3/helpers/interopRequireDefault");

var _Object$defineProperty = require("@babel/runtime-corejs3/core-js/object/define-property");

_Object$defineProperty(exports, "__esModule", {
  value: true
});

exports.handler = exports.builder = exports.description = exports.command = void 0;

var _terminalLink = _interopRequireDefault(require("terminal-link"));

var _lib = require("../../lib");

const command = 'introspect';
exports.command = command;
const description = 'Introspect your database and generate models in ./api/prisma/schema.prisma, overwriting existing models';
exports.description = description;

const builder = yargs => {
  yargs.option('verbose', {
    alias: 'v',
    default: true,
    description: 'Print more',
    type: 'boolean'
  }).epilogue(`Also see the ${(0, _terminalLink.default)('Redwood CLI Reference', 'https://redwoodjs.com/reference/command-line-interface#db-introspect')}`);
};

exports.builder = builder;

const handler = async ({
  verbose = true
}) => {
  return await (0, _lib.runCommandTask)([{
    title: 'Introspecting your database...',
    cmd: 'yarn prisma',
    args: ['introspect'],
    opts: {
      cwd: (0, _lib.getPaths)().api.db
    }
  }], {
    verbose
  });
};

exports.handler = handler;
