// @redwoodjs/cli/dist/commands/dbCommands/studio.js
"use strict";

var _interopRequireDefault = require("@babel/runtime-corejs3/helpers/interopRequireDefault");

var _Object$defineProperty = require("@babel/runtime-corejs3/core-js/object/define-property");

_Object$defineProperty(exports, "__esModule", {
  value: true
});

exports.handler = exports.builder = exports.description = exports.command = void 0;

var _path = _interopRequireDefault(require("path"));

var _fs = _interopRequireDefault(require("fs"));

var _terminalLink = _interopRequireDefault(require("terminal-link"));

var _lib = require("../../lib");

var _colors = _interopRequireDefault(require("../../lib/colors"));

const command = 'studio';
exports.command = command;
const description = 'Start Prisma Studio';
exports.description = description;

const builder = yargs => {
  yargs.epilogue(`Also see the ${(0, _terminalLink.default)('Redwood CLI Reference', 'https://redwoodjs.com/reference/command-line-interface#studio')}`);
};

exports.builder = builder;

const handler = async () => {
  // No schema, no studio.
  if (!_fs.default.existsSync((0, _lib.getPaths)().api.dbSchema)) {
    console.log(`${_colors.default.warning('[warning]')} cannot start Prisma Studio; schema missing (${_colors.default.info( // So we're not hard coding schema.prisma's relative location
    _path.default.relative((0, _lib.getPaths)().base, (0, _lib.getPaths)().api.dbSchema))}).`);
    return;
  }

  await (0, _lib.runCommandTask)([{
    title: 'Starting Prisma Studio...',
    cmd: 'yarn prisma',
    args: ['studio', '--experimental']
  }], {
    verbose: true
  });
};

exports.handler = handler;
