// @redwoodjs/web/dist/index.js
"use strict";

var _Object$keys = require("@babel/runtime-corejs3/core-js/object/keys");

var _forEachInstanceProperty = require("@babel/runtime-corejs3/core-js/instance/for-each");

var _context, _context2, _context3;

var _interopRequireDefault = require("@babel/runtime-corejs3/helpers/interopRequireDefault");

var _Object$defineProperty = require("@babel/runtime-corejs3/core-js/object/define-property");

_Object$defineProperty(exports, "__esModule", {
  value: true
});

var _exportNames = {
  gql: true,
  useQuery: true,
  useLazyQuery: true,
  useMutation: true,
  useSubscription: true,
  useApolloClient: true,
  FatalErrorBoundary: true,
  RedwoodProvider: true
};

_Object$defineProperty(exports, "gql", {
  enumerable: true,
  get: function get() {
    return _graphqlTag.default;
  }
});

_Object$defineProperty(exports, "useQuery", {
  enumerable: true,
  get: function get() {
    return _reactHooks.useQuery;
  }
});

_Object$defineProperty(exports, "useLazyQuery", {
  enumerable: true,
  get: function get() {
    return _reactHooks.useLazyQuery;
  }
});

_Object$defineProperty(exports, "useMutation", {
  enumerable: true,
  get: function get() {
    return _reactHooks.useMutation;
  }
});

_Object$defineProperty(exports, "useSubscription", {
  enumerable: true,
  get: function get() {
    return _reactHooks.useSubscription;
  }
});

_Object$defineProperty(exports, "useApolloClient", {
  enumerable: true,
  get: function get() {
    return _reactHooks.useApolloClient;
  }
});

_Object$defineProperty(exports, "FatalErrorBoundary", {
  enumerable: true,
  get: function get() {
    return _FatalErrorBoundary.default;
  }
});

_Object$defineProperty(exports, "RedwoodProvider", {
  enumerable: true,
  get: function get() {
    return _RedwoodProvider.default;
  }
});

require("./config");

var _graphqlTag = _interopRequireDefault(require("graphql-tag"));

var _reactHooks = require("@apollo/react-hooks");

var _FatalErrorBoundary = _interopRequireDefault(require("./components/FatalErrorBoundary"));

var _RedwoodProvider = _interopRequireDefault(require("./components/RedwoodProvider"));

var _graphql = require("./graphql");

_forEachInstanceProperty(_context = _Object$keys(_graphql)).call(_context, function (key) {
  if (key === "default" || key === "__esModule") return;
  if (Object.prototype.hasOwnProperty.call(_exportNames, key)) return;

  _Object$defineProperty(exports, key, {
    enumerable: true,
    get: function get() {
      return _graphql[key];
    }
  });
});

var _form = require("./form");

_forEachInstanceProperty(_context2 = _Object$keys(_form)).call(_context2, function (key) {
  if (key === "default" || key === "__esModule") return;
  if (Object.prototype.hasOwnProperty.call(_exportNames, key)) return;

  _Object$defineProperty(exports, key, {
    enumerable: true,
    get: function get() {
      return _form[key];
    }
  });
});

var _flash = require("./flash");

_forEachInstanceProperty(_context3 = _Object$keys(_flash)).call(_context3, function (key) {
  if (key === "default" || key === "__esModule") return;
  if (Object.prototype.hasOwnProperty.call(_exportNames, key)) return;

  _Object$defineProperty(exports, key, {
    enumerable: true,
    get: function get() {
      return _flash[key];
    }
  });
});
