// @redwoodjs/cli/dist/commands/storybook.js
"use strict";

var _interopRequireDefault = require("@babel/runtime-corejs3/helpers/interopRequireDefault");

var _Object$defineProperty = require("@babel/runtime-corejs3/core-js/object/define-property");

_Object$defineProperty(exports, "__esModule", {
  value: true
});

exports.handler = exports.description = exports.aliases = exports.command = void 0;

var _path = _interopRequireDefault(require("path"));

var _execa = _interopRequireDefault(require("execa"));

var _internal = require("@redwoodjs/internal");

const command = 'storybook';
exports.command = command;
const aliases = ['sb'];
exports.aliases = aliases;
const description = 'Launch Storybook, an isolated component development environment';
exports.description = description;

const handler = () => {
  const cmd = 'yarn start-storybook';
  const cwd = (0, _internal.getPaths)().web.base;

  const staticAssetsFolder = _path.default.join((0, _internal.getPaths)().web.base, 'public'); // Create the `MockServiceWorker.js` file.


  (0, _execa.default)(`yarn msw init ${staticAssetsFolder}`, undefined, {
    stdio: 'inherit',
    shell: true,
    cwd
  });
  (0, _execa.default)(cmd, ['--config-dir ../node_modules/@redwoodjs/core/config/storybook', '--port 7910', // this should be configurable?
  '--no-version-updates', // we'll handle upgrades
  '--ci', // do not open browser window.
  `--static-dir ${staticAssetsFolder}`], {
    stdio: 'inherit',
    shell: true,
    cwd
  });
};

exports.handler = handler;
