// @redwoodjs/structure/dist/index.js
"use strict";

var _Object$defineProperty = require("@babel/runtime-corejs3/core-js/object/define-property");

_Object$defineProperty(exports, "__esModule", {
  value: true
});

exports.getProject = getProject;
exports.printDiagnostics = printDiagnostics;

_Object$defineProperty(exports, "DefaultHost", {
  enumerable: true,
  get: function () {
    return _ide.DefaultHost;
  }
});

_Object$defineProperty(exports, "Host", {
  enumerable: true,
  get: function () {
    return _ide.Host;
  }
});

_Object$defineProperty(exports, "RWProject", {
  enumerable: true,
  get: function () {
    return _model.RWProject;
  }
});

_Object$defineProperty(exports, "DiagnosticSeverity", {
  enumerable: true,
  get: function () {
    return _vscodeLanguageserverTypes2.DiagnosticSeverity;
  }
});

_Object$defineProperty(exports, "URL_file", {
  enumerable: true,
  get: function () {
    return _URL.URL_file;
  }
});

var _ide = require("./ide");

var _model = require("./model");

var _vscodeLanguageserverTypes = require("./x/vscode-languageserver-types");

var _vscodeLanguageserverTypes2 = require("vscode-languageserver-types");

var _URL = require("./x/URL");

function getProject(projectRoot, host = new _ide.DefaultHost()) {
  return new _model.RWProject({
    projectRoot,
    host
  });
}

async function printDiagnostics(projectRoot, opts) {
  const project = getProject(projectRoot);
  const formatOpts = {
    cwd: projectRoot,
    ...opts
  };

  try {
    for (const d of await project.collectDiagnostics()) {
      const str = (0, _vscodeLanguageserverTypes.ExtendedDiagnostic_format)(d, formatOpts);
      console.log(str);
    }
  } catch (e) {
    console.log('runtime error: ' + e.message);
  }
}
